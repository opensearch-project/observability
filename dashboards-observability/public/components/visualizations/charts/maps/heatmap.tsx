/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useMemo } from 'react';
import { colorPalette } from '@elastic/eui';
import { forEach, has, isEmpty, isEqual, uniq } from 'lodash';
import Plotly from 'plotly.js-dist';
import {
  HEATMAP_PALETTE_COLOR,
  HEATMAP_SINGLE_COLOR,
  OPACITY,
  SINGLE_COLOR_PALETTE,
} from '../../../../../common/constants/colors';
import {
  hexToRgb,
  lightenColor,
  getPropName,
} from '../../../../components/event_analytics/utils/utils';
import { IVisualizationContainerProps } from '../../../../../common/types/explorer';
import { VisCanvassPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components/vis_canvass_placeholder';
import { Plt } from '../../plotly/plot';
import { AGGREGATIONS, GROUPBY } from '../../../../../common/constants/explorer';
import { PLOT_MARGIN } from '../../../../../common/constants/shared';
import { getCompleteTimespanKey } from '../../../visualizations/charts/shared/common';
import { removeBacktick } from '../../../../../common/utils';

export const HeatMap = ({ visualizations, layout, config }: any) => {
  const {
    data: {
      rawVizData: {
        data: queriedVizData,
        metadata: { fields },
      },
      userConfigs: {
        dataConfig: {
          chartStyles = {},
          span = {},
          legend = {},
          tooltipOptions = {},
          panelOptions = {},
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
        } = {},
        layoutConfig = {},
      } = {},
    } = {},
    vis: { icontype },
  }: IVisualizationContainerProps = visualizations;

  const backtickRemovedVisData = {};
  forEach(queriedVizData, (value, key) => {
    backtickRemovedVisData[removeBacktick(key)] = value;
  });

  const combinedDemensions = [
    ...(!isEmpty(span) ? [getCompleteTimespanKey(span)] : []),
    ...dimensions,
  ];
  if (!isEqual(combinedDemensions.length, 2) || !isEqual(series.length, 1))
    return <VisCanvassPlaceholder message="Invalid data, heatmap can only have exact 2 dimensions and 1 series" icon={icontype} />;
  const xaxisField = { ...combinedDemensions[0] };
  const yaxisField = { ...combinedDemensions[1] };
  const zMetrics = { ...series[0] };
  if (
    isEmpty(xaxisField) ||
    isEmpty(yaxisField) ||
    isEmpty(zMetrics) ||
    isEmpty(backtickRemovedVisData[removeBacktick(xaxisField.label)]) ||
    isEmpty(backtickRemovedVisData[removeBacktick(yaxisField.label)])
  )
    return <VisCanvassPlaceholder message="Invalid data" icon={icontype} />;

  const uniqueYaxis = uniq(backtickRemovedVisData[removeBacktick(yaxisField.label)]);
  const uniqueXaxis = uniq(backtickRemovedVisData[removeBacktick(xaxisField.label)]);
  const uniqueYaxisLength = uniqueYaxis.length;
  const uniqueXaxisLength = uniqueXaxis.length;
  const tooltipMode =
    tooltipOptions.tooltipMode !== undefined ? tooltipOptions.tooltipMode : 'show';
  const tooltipText = tooltipOptions.tooltipText !== undefined ? tooltipOptions.tooltipText : 'all';

  const colorField = chartStyles
    ? chartStyles.colorMode && chartStyles.colorMode[0].name === OPACITY
      ? chartStyles.color ?? HEATMAP_SINGLE_COLOR
      : chartStyles.scheme ?? HEATMAP_PALETTE_COLOR
    : HEATMAP_PALETTE_COLOR;
  const showColorscale = legend.showLegend ?? 'show';

  const traceColor: any = [];
  if (colorField.name === SINGLE_COLOR_PALETTE) {
    const colorsArray = colorPalette([lightenColor(colorField.color, 50), colorField.color], 10);
    colorsArray.map((hexCode, index) => {
      traceColor.push([
        (index !== colorsArray.length - 1 ? index : 10) / 10,
        hexToRgb(hexCode, 1, false),
      ]);
    });
  }

  const calculatedHeapMapZaxis: Plotly.Data[] = useMemo(() => {
    const heapMapZaxis = [];
    const buckets = {};

    // maps bukcets to metrics
    for (let i = 0; i < backtickRemovedVisData[removeBacktick(xaxisField.label)].length; i++) {
      buckets[`${backtickRemovedVisData[removeBacktick(xaxisField.label)][i]},${backtickRemovedVisData[removeBacktick(yaxisField.label)][i]}`] =
      backtickRemovedVisData[getPropName(zMetrics)][i];
    }

    // initialize empty 2 dimensional array, inner loop for each xaxis field, outer loop for yaxis
    for (let i = 0; i < uniqueYaxisLength; i++) {
      const innerBuckets = [];
      for (let j = 0; j < uniqueXaxisLength; j++) {
        innerBuckets.push(null);
      }
      heapMapZaxis.push(innerBuckets);
    }

    // fill in each bucket
    for (let i = 0; i < uniqueYaxisLength; i++) {
      for (let j = 0; j < uniqueXaxisLength; j++) {
        if (has(buckets, `${uniqueXaxis[j]},${uniqueYaxis[i]}`)) {
          heapMapZaxis[i][j] = buckets[`${uniqueXaxis[j]},${uniqueYaxis[i]}`];
        }
      }
    }

    return heapMapZaxis;
  }, [
    backtickRemovedVisData,
    uniqueYaxis,
    uniqueXaxis,
    uniqueYaxisLength,
    uniqueXaxisLength,
    xaxisField,
    yaxisField,
    zMetrics,
  ]);

  const heapMapData = [
    {
      z: calculatedHeapMapZaxis,
      x: uniqueXaxis,
      y: uniqueYaxis,
      hoverinfo: tooltipMode === 'hidden' ? 'none' : tooltipText,
      colorscale: colorField.name === SINGLE_COLOR_PALETTE ? traceColor : colorField.name,
      type: 'heatmap',
      showscale: showColorscale === 'show',
    },
  ];

  layout.yaxis = { autosize: true, automargin: true };
  const mergedLayout = {
    ...layout,
    ...(layoutConfig.layout && layoutConfig.layout),
    title: panelOptions.title || layoutConfig.layout?.title || '',
    margin: PLOT_MARGIN,
  };

  const mergedConfigs = useMemo(
    () => ({
      ...config,
      ...(layoutConfig.config && layoutConfig.config),
    }),
    [config, layoutConfig.config]
  );

  return <Plt data={heapMapData} layout={mergedLayout} config={mergedConfigs} />;
};
