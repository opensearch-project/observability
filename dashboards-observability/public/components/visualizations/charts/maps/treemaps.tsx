/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { isEmpty, isEqual, uniq } from 'lodash';
import React, { useMemo } from 'react';

import {
  DEFAULT_PALETTE,
  MULTI_COLOR_PALETTE,
  SINGLE_COLOR_PALETTE,
} from '../../../../../common/constants/colors';
import { AGGREGATIONS, GROUPBY } from '../../../../../common/constants/explorer';
import { DEFAULT_CHART_STYLES, PLOT_MARGIN } from '../../../../../common/constants/shared';
import { IVisualizationContainerProps } from '../../../../../common/types/explorer';
import { EmptyPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components/empty_placeholder';
import { Plt } from '../../plotly/plot';

export const TreeMap = ({ visualizations, layout, config }: any) => {
  const { DefaultSortSectors } = DEFAULT_CHART_STYLES;
  const {
    data: {
      defaultAxes,
      indexFields,
      query,
      rawVizData: {
        data: queriedVizData,
        metadata: { fields },
      },
      userConfigs: {
        dataConfig: {
          chartStyles = {},
          legend = {},
          tooltipOptions = {},
          panelOptions = {},
          treemapOptions = {},
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
        },
        layoutConfig = {},
      },
    },
    vis: { icontype },
  }: IVisualizationContainerProps = visualizations;

  const childField =
    dimensions && dimensions[0]?.childField ? dimensions[0]?.childField : fields[fields.length - 1];
  const parentFields = dimensions && dimensions[0]?.parentFields ? dimensions[0]?.parentFields : [];
  const tooltipMode =
    tooltipOptions.tooltipMode !== undefined ? tooltipOptions.tooltipMode : 'show';
  const tooltipText = tooltipOptions.tooltipText !== undefined ? tooltipOptions.tooltipText : 'all';
  const valueField = series && series[0]?.valueField ? series[0]?.valueField : fields[0];
  const colorField =
    chartStyles && chartStyles.colorTheme ? chartStyles.colorTheme : { name: DEFAULT_PALETTE };

  const tilingAlgorithm =
    treemapOptions && treemapOptions.tilingAlgorithm && !isEmpty(treemapOptions.tilingAlgorithm)
      ? treemapOptions.tilingAlgorithm[0]
      : 'squarify';

  const sortSectorsField = treemapOptions.sort_sectors || DefaultSortSectors;
  const showColorscale = legend.showLegend ?? 'show';

  const areParentFieldsInvalid =
    new Set([...parentFields.map((field) => field.name)]).size !== parentFields.length ||
    parentFields.some(
      (field) => isEmpty(queriedVizData[field.name]) || isEqual(childField.name, field.name)
    );

  if (
    isEmpty(queriedVizData[childField.name]) ||
    isEmpty(queriedVizData[valueField.name]) ||
    areParentFieldsInvalid
  )
    return <EmptyPlaceholder icon={icontype} />;

  const [treemapData, mergedLayout] = useMemo(() => {
    let labelsArray: string[] = [];
    let parentsArray: string[] = [];
    let valuesArray: number[] = [];
    let colorsArray: string[] = [];

    if (parentFields.length === 0) {
      labelsArray = [...queriedVizData[childField.name]];
      parentsArray = [...Array(labelsArray.length).fill('')];
      valuesArray = [...queriedVizData[valueField.name]];
      if (colorField.name === MULTI_COLOR_PALETTE) {
        colorsArray = [
          ...Array(queriedVizData[childField.name].length).fill(colorField.childColor),
        ];
      }
    } else {
      let currentLevel = parentFields.length - 1;
      let lastParentField = {};
      parentFields
        .slice(0)
        .reverse()
        .map((field, i) => {
          const uniqueParents = uniq(queriedVizData[field.name]) as string[];
          labelsArray = [...labelsArray, ...uniqueParents];
          if (i === 0) {
            parentsArray = [...Array(uniqueParents.length).fill('')];
            valuesArray = [...Array(uniqueParents.length).fill(0)];
            colorsArray =
              colorField.name === MULTI_COLOR_PALETTE
                ? [
                    ...Array(uniqueParents.length).fill(
                      colorField.parentColors[currentLevel] ?? '#000000'
                    ),
                  ]
                : [];
          } else {
            const currentParentIndices = uniqueParents.map((parent) =>
              queriedVizData[field.name].findIndex((index) => index === parent)
            );
            const lastParents = currentParentIndices.map(
              (index) => queriedVizData[lastParentField.name][index]
            );
            parentsArray = [...parentsArray, ...lastParents];
            valuesArray = [...valuesArray, ...Array(lastParents.length).fill(0)];
            colorsArray =
              colorField.name === MULTI_COLOR_PALETTE
                ? [
                    ...colorsArray,
                    ...Array(lastParents.length).fill(
                      colorField.parentColors[currentLevel] ?? '#000000'
                    ),
                  ]
                : [];
          }
          currentLevel = currentLevel - 1;
          lastParentField = field;
        });

      labelsArray = [...labelsArray, ...queriedVizData[childField.name]];
      valuesArray = [...valuesArray, ...queriedVizData[valueField.name]];
      parentsArray = [...parentsArray, ...queriedVizData[lastParentField.name]];
      colorsArray =
        colorField.name === MULTI_COLOR_PALETTE
          ? [
              ...colorsArray,
              ...Array(queriedVizData[childField.name].length).fill(colorField.childColor),
            ]
          : [];
    }

    if (colorField.name === SINGLE_COLOR_PALETTE) {
      colorsArray = [...Array(valuesArray.length).fill(colorField.childColor)];
    }

    const markerColors =
      colorField.name === MULTI_COLOR_PALETTE
        ? { colors: colorsArray }
        : ![DEFAULT_PALETTE, SINGLE_COLOR_PALETTE].includes(colorField.name)
        ? {
            colorscale: colorField.name,
            colorbar: {
              len: 1,
            },
            showscale: showColorscale === 'show',
          }
        : {};

    const colorway = colorField.name === SINGLE_COLOR_PALETTE ? colorsArray : {};

    const mapLayout = {
      ...layout,
      ...(layoutConfig.layout && layoutConfig.layout),
      title: panelOptions.title || layoutConfig.layout?.title || '',
      treemapcolorway: colorway,
      margin: PLOT_MARGIN,
    };

    const mapData = [
      {
        type: 'treemap',
        labels: labelsArray,
        parents: parentsArray,
        values: valuesArray,
        hoverinfo: tooltipMode === 'hidden' ? 'none' : tooltipText,
        textinfo: 'label+value+percent parent+percent entry',
        tiling: {
          packing: tilingAlgorithm.value,
        },
        marker: markerColors,
        sort: sortSectorsField === DefaultSortSectors,
      },
    ];

    return [mapData, mapLayout];
  }, [
    queriedVizData,
    childField,
    valueField,
    parentFields,
    colorField,
    tilingAlgorithm,
    layoutConfig,
  ]);

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };

  return <Plt data={treemapData} layout={mergedLayout} config={mergedConfigs} />;
};
