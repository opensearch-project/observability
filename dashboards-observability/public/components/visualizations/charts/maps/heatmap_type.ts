/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { HeatMap } from './heatmap';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartPie } from '../../assets/chart_pie';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import {
  ConfigValueOptions,
  HeatmapColorPalettePicker,
  ConfigChartOptions,
  PanelItem,
  SingleColorPicker,
  ConfigLegend,
} from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import {
  COLOR_PALETTES,
  HEATMAP_SINGLE_COLOR,
  HEATMAP_PALETTE_COLOR,
} from '../../../../../common/constants/colors';
import { fetchConfigObject } from '../../../../components/event_analytics/utils/utils';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();

export const createMapsVisDefinition = () => ({
  name: 'heatmap',
  type: 'heatmap',
  id: 'heatmap',
  label: 'Heatmap',
  fulllabel: 'Hubble',
  icontype: 'heatmap',
  category: VIS_CATEGORY.BASICS,
  selection: {
    dataLoss: 'nothing',
  },
  icon: LensIconChartPie,
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          fetchConfigObject('Tooltip', {
            options: [
              { name: 'All', id: 'all' },
              { name: 'Dim 1', id: 'x' },
              { name: 'Dim 2', id: 'y' },
              { name: 'Metrics', id: 'z' },
            ],
            defaultSelections: [{ name: 'All', id: 'all' }],
          }),
          {
            id: 'legend',
            name: 'Legend',
            editor: ConfigLegend,
            mapTo: 'legend',
            schemas: [
              {
                name: 'Show colorscale',
                mapTo: 'showLegend',
                component: null,
                props: {
                  options: [
                    { name: 'Show', id: 'show' },
                    { name: 'Hidden', id: 'hidden' },
                  ],
                  defaultSelections: [{ name: 'Show', id: 'show' }],
                },
              },
            ],
          },
          {
            id: 'chart_styles',
            name: 'Chart styles',
            editor: ConfigChartOptions,
            mapTo: 'chartStyles',
            schemas: [
              {
                name: 'Color mode',
                component: PanelItem,
                mapTo: 'colorMode',
                eleType: 'list',
                isSingleSelection: true,
                options: [
                  { name: 'spectrum', label: 'spectrum', value: 'spectrum' },
                  { name: 'opacity', label: 'opacity', value: 'opacity' },
                ],
                defaultState: [{ name: 'spectrum', label: 'spectrum', value: 'spectrum' }],
                props: {
                  isClearable: false,
                },
              },
              {
                name: 'Scheme',
                component: HeatmapColorPalettePicker,
                mapTo: 'scheme',
                eleType: 'palettePicker',
                options: COLOR_PALETTES.filter((color) => color.type !== 'text'),
                defaultState: HEATMAP_PALETTE_COLOR,
              },
              {
                name: 'Color',
                component: SingleColorPicker,
                mapTo: 'color',
                eleType: 'singleColorPicker',
                defaultState: HEATMAP_SINGLE_COLOR,
              },
            ],
          },
        ],
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
      plot_bgcolor: 'rgba(0, 0, 0, 0)',
      paper_bgcolor: 'rgba(0, 0, 0, 0)',
      margin: { left: 60 },
    },
    config: {
      ...sharedConfigs.config,
    },
  },
  component: HeatMap,
});
