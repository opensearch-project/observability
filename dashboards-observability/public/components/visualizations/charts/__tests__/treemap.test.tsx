/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { TreeMap } from '../maps/treemaps';
import {
  LAYOUT_CONFIG,
  TEST_VISUALIZATIONS_DATA,
} from '../../../../../test/event_analytics_constants';

describe('Treemap component', () => {
  configure({ adapter: new Adapter() });

  it('Renders treemap component', async () => {
    const wrapper = mount(
      <TreeMap visualizations={TEST_VISUALIZATIONS_DATA} layout={LAYOUT_CONFIG} config={{}} />
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
