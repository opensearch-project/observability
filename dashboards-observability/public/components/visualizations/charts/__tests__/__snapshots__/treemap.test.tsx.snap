// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Treemap component Renders treemap component 1`] = `
<TreeMap
  config={Object {}}
  layout={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  visualizations={
    Object {
      "data": Object {
        "appData": Object {
          "fromApp": false,
        },
        "defaultAxes": Object {},
        "indexFields": Object {
          "availableFields": Array [
            Object {
              "name": "agent",
              "type": "string",
            },
            Object {
              "name": "bytes",
              "type": "long",
            },
            Object {
              "name": "clientip",
              "type": "ip",
            },
            Object {
              "name": "event",
              "type": "struct",
            },
            Object {
              "name": "extension",
              "type": "string",
            },
            Object {
              "name": "geo",
              "type": "struct",
            },
            Object {
              "name": "host",
              "type": "string",
            },
            Object {
              "name": "index",
              "type": "string",
            },
            Object {
              "name": "ip",
              "type": "ip",
            },
            Object {
              "name": "machine",
              "type": "struct",
            },
            Object {
              "name": "memory",
              "type": "double",
            },
            Object {
              "name": "message",
              "type": "string",
            },
            Object {
              "name": "phpmemory",
              "type": "long",
            },
            Object {
              "name": "referer",
              "type": "string",
            },
            Object {
              "name": "request",
              "type": "string",
            },
            Object {
              "name": "response",
              "type": "string",
            },
            Object {
              "name": "tags",
              "type": "string",
            },
            Object {
              "name": "timestamp",
              "type": "timestamp",
            },
            Object {
              "name": "url",
              "type": "string",
            },
            Object {
              "name": "utc_time",
              "type": "timestamp",
            },
          ],
          "queriedFields": Array [
            Object {
              "name": "double_per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "host",
              "type": "text",
            },
            Object {
              "name": "ip_count",
              "type": "integer",
            },
            Object {
              "name": "per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "resp_code",
              "type": "text",
            },
            Object {
              "name": "sum_bytes",
              "type": "long",
            },
          ],
          "selectedFields": Array [],
          "unselectedFields": Array [],
        },
        "query": Object {},
        "rawVizData": Object {
          "data": Object {
            "count()": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
            "tags": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
          },
          "jsonData": Array [
            Object {
              "count()": 154,
              "tags": "error",
            },
            Object {
              "count()": 1753,
              "tags": "info",
            },
            Object {
              "count()": 116,
              "tags": "login",
            },
            Object {
              "count()": 468,
              "tags": "security",
            },
            Object {
              "count()": 1964,
              "tags": "success",
            },
            Object {
              "count()": 219,
              "tags": "warning",
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count()",
                "type": "integer",
              },
              Object {
                "name": "tags",
                "type": "text",
              },
            ],
            "size": 6,
            "status": 200,
          },
        },
        "userConfigs": Object {
          "dataConfig": Object {
            "valueOptions": Object {
              "dimensions": Array [
                Object {
                  "label": "tags",
                  "name": "tags",
                  "type": "text",
                },
              ],
              "metrics": Array [
                Object {
                  "label": "count()",
                  "name": "count()",
                  "side": "left",
                  "type": "integer",
                },
              ],
            },
          },
        },
      },
      "vis": Object {
        "barwidth": 0.97,
        "category": "Visualizations",
        "categoryaxis": "xaxis",
        "component": [Function],
        "editorconfig": Object {
          "panelTabs": Array [
            Object {
              "editor": [Function],
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Style",
              "sections": Array [
                Object {
                  "editor": [Function],
                  "id": "tooltip_options",
                  "mapTo": "tooltipOptions",
                  "name": "Tooltip options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "mapTo": "tooltipMode",
                      "name": "Tooltip mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "show",
                            "name": "Show",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "show",
                            "name": "Show",
                          },
                          Object {
                            "id": "hidden",
                            "name": "Hidden",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "mapTo": "tooltipText",
                      "name": "Tooltip text",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "all",
                            "name": "All",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "all",
                            "name": "All",
                          },
                          Object {
                            "id": "x",
                            "name": "Dimension",
                          },
                          Object {
                            "id": "y",
                            "name": "Series",
                          },
                        ],
                      },
                    },
                  ],
                },
                Object {
                  "editor": [Function],
                  "id": "legend",
                  "mapTo": "legend",
                  "name": "Legend",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "mapTo": "showLegend",
                      "name": "Show legend",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "show",
                            "name": "Show",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "show",
                            "name": "Show",
                          },
                          Object {
                            "id": "hidden",
                            "name": "Hidden",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "mapTo": "position",
                      "name": "Position",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "v",
                            "name": "Right",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "v",
                            "name": "Right",
                          },
                          Object {
                            "id": "h",
                            "name": "Bottom",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "legendSize",
                      "name": "Legend size",
                      "title": "Legend size",
                    },
                  ],
                },
                Object {
                  "editor": [Function],
                  "id": "chart_styles",
                  "mapTo": "chartStyles",
                  "name": "Chart styles",
                  "schemas": Array [
                    Object {
                      "component": [Function],
                      "eleType": "buttons",
                      "mapTo": "mode",
                      "name": "Mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "group",
                            "name": "Group",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "group",
                            "name": "Group",
                          },
                          Object {
                            "id": "stack",
                            "name": "Stack",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "labelSize",
                      "name": "Label size",
                    },
                    Object {
                      "component": [Function],
                      "defaultState": 0,
                      "eleType": "slider",
                      "mapTo": "rotateBarLabels",
                      "name": "Rotate bar labels",
                      "props": Object {
                        "max": 90,
                        "min": -90,
                        "showTicks": true,
                        "ticks": Array [
                          Object {
                            "label": "-90°",
                            "value": -90,
                          },
                          Object {
                            "label": "-45°",
                            "value": -45,
                          },
                          Object {
                            "label": "0°",
                            "value": 0,
                          },
                          Object {
                            "label": "45°",
                            "value": 45,
                          },
                          Object {
                            "label": "90°",
                            "value": 90,
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "defaultState": 0.7,
                      "eleType": "slider",
                      "mapTo": "groupWidth",
                      "name": "Group width",
                      "props": Object {
                        "max": 1,
                        "step": 0.01,
                      },
                    },
                    Object {
                      "component": [Function],
                      "defaultState": 0.97,
                      "eleType": "slider",
                      "mapTo": "barWidth",
                      "name": "Bar width",
                      "props": Object {
                        "max": 1,
                        "step": 0.01,
                      },
                    },
                    Object {
                      "component": [Function],
                      "defaultState": 2,
                      "eleType": "slider",
                      "mapTo": "lineWidth",
                      "name": "Line width",
                      "props": Object {
                        "max": 10,
                      },
                    },
                    Object {
                      "component": [Function],
                      "defaultState": 70,
                      "eleType": "slider",
                      "mapTo": "fillOpacity",
                      "name": "Fill opacity",
                      "props": Object {
                        "max": 100,
                      },
                    },
                  ],
                },
              ],
            },
            Object {
              "editor": [Function],
              "id": "availability-panel",
              "mapTo": "availabilityConfig",
              "name": "Availability",
            },
          ],
        },
        "fillopacity": 70,
        "fulllabel": "Vertical bar",
        "groupwidth": 0.7,
        "icon": [Function],
        "icontype": "visBarVerticalStacked",
        "id": "bar",
        "label": "Vertical bar",
        "labelangle": 0,
        "legendposition": "v",
        "linewidth": 2,
        "mode": "group",
        "name": "bar",
        "orientation": "v",
        "selection": Object {
          "dataLoss": "nothing",
        },
        "seriesaxis": "yaxis",
        "showlegend": "show",
        "type": "bar",
        "visconfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "isUniColor": false,
          "layout": Object {
            "height": 1180,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "showlegend": true,
          },
        },
      },
    }
  }
>
  <Plt
    config={Object {}}
    data={
      Array [
        Object {
          "hoverinfo": "all",
          "labels": Array [
            "error",
            "info",
            "login",
            "security",
            "success",
            "warning",
          ],
          "marker": Object {},
          "parents": Array [
            "",
            "",
            "",
            "",
            "",
            "",
          ],
          "sort": true,
          "textinfo": "label+value+percent parent+percent entry",
          "tiling": Object {
            "packing": undefined,
          },
          "type": "treemap",
          "values": Array [
            154,
            1753,
            116,
            468,
            1964,
            219,
          ],
        },
      ]
    }
    layout={
      Object {
        "colorway": Array [
          "#8C55A3",
        ],
        "height": 220,
        "margin": Object {
          "b": 30,
          "l": 30,
          "pad": 4,
          "r": 5,
          "t": 50,
        },
        "showlegend": true,
        "title": "",
        "treemapcolorway": Object {},
      }
    }
  >
    <PlotlyComponent
      config={
        Object {
          "displayModeBar": false,
        }
      }
      data={
        Array [
          Object {
            "hoverinfo": "all",
            "labels": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
            "marker": Object {},
            "parents": Array [
              "",
              "",
              "",
              "",
              "",
              "",
            ],
            "sort": true,
            "textinfo": "label+value+percent parent+percent entry",
            "tiling": Object {
              "packing": undefined,
            },
            "type": "treemap",
            "values": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
          },
        ]
      }
      debug={false}
      divId="explorerPlotComponent"
      layout={
        Object {
          "autosize": true,
          "barmode": "stack",
          "colorway": Array [
            "#8C55A3",
          ],
          "height": 220,
          "hovermode": "closest",
          "legend": Object {
            "orientation": "h",
            "traceorder": "normal",
          },
          "margin": Object {
            "b": 30,
            "l": 30,
            "pad": 4,
            "r": 5,
            "t": 50,
          },
          "showlegend": true,
          "title": "",
          "treemapcolorway": Object {},
          "xaxis": Object {
            "automargin": true,
            "rangemode": "normal",
            "showgrid": true,
            "zeroline": false,
          },
          "yaxis": Object {
            "rangemode": "normal",
            "showgrid": true,
            "zeroline": false,
          },
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useResizeHandler={true}
    >
      <div
        id="explorerPlotComponent"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </PlotlyComponent>
  </Plt>
</TreeMap>
`;
